AzureMonitor: Traces - Testing

aangelisc opened the pull request:
This PR adds testing to the existing Traces work. This includes frontend, backend, and e2e testing. 

- Fix bug where the correct `queryType` wasn't selected, preventing template variables from being interpolated.
- Fix/update existing tests.
- Ensure test selectors are consistent.
- Added `TracesQueryEditor` test.
- Added `FormatAsField` test.
- Added `TraceTypeField` test.
- Added `Filters` test.
- Fixed multi-selection on `Filters`.
- Updated backend datasource test.
- Updated backend response-table-frame test.
- Updated backend macros test.
- Added relevant test data.
- Added generation for test data using `CheckGoldenJSONResponse` util.
- Updated all tests and removed old `expectedFrame` properties.

Part of #63117, #63127, #63124, #63118, #63121, #63119, #63129
Fixes #63123

aangelisc committed:
Select ds for template variable interpolation

aangelisc committed:
Update az logs ds tests

- Add templateVariables test
- Add filter test
- Update mock
- Remove anys

aangelisc committed:
Update QueryEditor test

- Update mocks with timeSrv for log analytics datasource
- Fix query mock
- Use appropriate and consistent selectors

aangelisc committed:
Add TracesQueryEditor test

- Update resourcePickerRows mock to include app insights resources
- Remove comments and extra new line

aangelisc committed:
Add FormatAsField test

- Remove unneeded condition

aangelisc committed:
Update resourcePicker utils test

aangelisc committed:
Merge branch 'azure-monitor-traces' into azure-traces-testing

aangelisc committed:
Don't hide selected options in filters

aangelisc committed:
Fix multi-selection on filters

aangelisc committed:
Add TraceTypeField test

- Add test file
- Update selectors (remove copy/paste mistake)
- Update placeholder text for select and add label

aangelisc committed:
Add basic filters test

aangelisc committed:
Begin filters test

aangelisc committed:
Update filters test

aangelisc committed:
Add final tests and simplify/generalise addFilter helper

aangelisc commented:
@JoaoSilvaGrafana I had to add another property to get my tests to work because of the way the `react-dom` works in tests (all DOMs are touch capable by definition which breaks my tests). Let me know if this is a problem, I'm happy to work with you on reducing the number of props if needed.

joshhunt commented on .betterer.results:
```diff
@@ -3816,8 +3816,7 @@ exports[`better eslint`] = {
     "public/app/plugins/datasource/azuremonitor/azure_log_analytics/azure_log_analytics_datasource.test.ts:5381": [
       [0, 0, 0, "Unexpected any. Specify a different type.", "0"],
       [0, 0, 0, "Unexpected any. Specify a different type.", "1"],
-      [0, 0, 0, "Unexpected any. Specify a different type.", "2"],
-      [0, 0, 0, "Unexpected any. Specify a different type.", "3"]
```
🚀

aangelisc committed:
Minor update to datasource test

aangelisc committed:
Update macros test

aangelisc committed:
Update selectors in tests

aangelisc committed:
Merge branch 'azure-monitor-traces' into azure-traces-testing

aangelisc committed:
Add response-table-frame tests

aangelisc committed:
Add datasource tests

- Use sorting where JSON models are inconsistent
- Update filters clause
- Dedupe tags
- Correct operationId conditions

aangelisc committed:
Merge branch 'azure-monitor-traces' into azure-traces-testing

aangelisc committed:
Don't set a default value for blurInputOnSelect

aangelisc committed:
Simplify datasource test

JoaoSilvaGrafana commented:
> @JoaoSilvaGrafana I had to add another property to get my tests to work because of the way the `react-dom` works in tests (all DOMs are touch capable by definition which breaks my tests). Let me know if this is a problem, I'm happy to work with you on reducing the number of props if needed.

As discussed, not ideal to keep adding props but we will revisit this on the Select component rework so its ok for now 👍

asimpson reviewed (approved):
Whew, nicely done on this! :trophy:

bossinc commented on pkg/tsdb/azuremonitor/loganalytics/azure-response-table-frame_test.go:
```diff
@@ -193,7 +194,418 @@ func TestLogTableToFrame(t *testing.T) {
+		{
+			name:     "multi trace",
+			testFile: "traces/1-traces-multiple-table.json",
+			expectedFrame: func() *data.Frame {
```
This looks like a good use case for `CheckGoldenJSONResponse`. Here is an example of a helper function. https://github.com/grafana/github-datasource/blob/main/pkg/testutil/frames.go#L21

aangelisc replied:
I'll have a look at this 👀

aangelisc replied:
Thank you @bossinc for the great recommendation! That's a bunch of redundant code removed 😊

bossinc reviewed (commented):
so much testing! i love it

aangelisc committed:
Update to use CheckGoldenJSON utils

- Update with generated frame files
- Remove redundant expected frame code
- Update all usages

aangelisc committed:
Merge branch 'azure-monitor-traces' into azure-traces-testing

# Conflicts:
#	pkg/tsdb/azuremonitor/metrics/azuremonitor-datasource_test.go

aangelisc committed:
Fix lint
